File: iBurn/src/main/java/com/gaiagps/iburn/service/DataUpdateService.java
package com.gaiagps.iburn.service;

import android.content.Context;
import android.transition.Explode;

import com.gaiagps.iburn.api.IBurnService;
import com.google.android.gms.gcm.GcmNetworkManager;
import com.google.android.gms.gcm.GcmTaskService;
import com.google.android.gms.gcm.PeriodicTask;
import com.google.android.gms.gcm.Task;
import com.google.android.gms.gcm.TaskParams;

import net.hockeyapp.android.ExceptionHandler;

import timber.log.Timber;

/**
 * This service gets awoken by the Google Cloud Messaging Network Manager
 * and allows us to perform our update task at opportune times.
 * <p>
 * Created by davidbrodsky on 7/2/15.
 */
public class DataUpdateService extends GcmTaskService {

    public static final String AUTO_UPDATE_TASK_NAME = "iburn-auto-update";

    public static void scheduleAutoUpdate(Context context) {
        long periodSecs = 60 * 60 * 24;     // Auto-update should be performed no more than once per 24 hours

        PeriodicTask dailyUpdate = new PeriodicTask.Builder()
                .setService(DataUpdateService.class)
                .setPeriod(periodSecs)
                .setTag(AUTO_UPDATE_TASK_NAME)
                .setPersisted(true)
                .setRequiredNetwork(Task.NETWORK_STATE_CONNECTED)
                .setRequiresCharging(false)
                .setUpdateCurrent(true)
                .build();

        GcmNetworkManager.getInstance(context).schedule(dailyUpdate);
        Timber.d("Scheduled auto-update");
    }

    /**
     * Called on application or Google Play Services update
     */
    @Override
    public void onInitializeTasks() {
        scheduleAutoUpdate(getApplicationContext());
    }

    @Override
    public int onRunTask(TaskParams taskParams) {
        try {
            if (taskParams.getTag().equals(AUTO_UPDATE_TASK_NAME)) {
                Timber.d("GCM invoked update task");
                IBurnService service = new IBurnService(getApplicationContext());
                boolean success = service.updateData().singleOrDefault(true).toBlocking().single();
                Timber.d("GCM invoked task finished with success %b", success);
                return success ? GcmNetworkManager.RESULT_SUCCESS : GcmNetworkManager.RESULT_RESCHEDULE;
            }
            Timber.w("Unknown task (%s) invoked", taskParams.getTag());
            return GcmNetworkManager.RESULT_FAILURE;
        } catch (Exception e) {
            ExceptionHandler.saveException(e, null);
            Timber.e(e, "GCM task failed: %s", e.getClass().getSimpleName());
            return GcmNetworkManager.RESULT_RESCHEDULE;
        }
    }
}

File: iBurn/src/main/java/com/gaiagps/iburn/service/DataUpdateService.kt
package com.gaiagps.iburn.service

import android.content.Context
import androidx.work.Constraints
import androidx.work.NetworkType
import androidx.work.PeriodicWorkRequestBuilder
import androidx.work.WorkManager
import androidx.work.Worker
import androidx.work.WorkerParameters
import com.gaiagps.iburn.BuildConfig
import com.gaiagps.iburn.api.IBurnService
import com.gaiagps.iburn.database.DataProvider
import kotlinx.coroutines.runBlocking
import timber.log.Timber
import java.util.concurrent.TimeUnit


class DataUpdateService(context: Context, workerParams: WorkerParameters) :
    Worker(context, workerParams) {

    companion object {
        fun scheduleAutoUpdate(context: Context) {
            val periodHrs = 24L  // Auto-update should be performed no more than once per 24 hours

            val constraints = Constraints.Builder()
                .setRequiredNetworkType(NetworkType.CONNECTED)
                .setRequiresCharging(false)
                .build()

            val request = PeriodicWorkRequestBuilder<DataUpdateService>(periodHrs, TimeUnit.HOURS)
                .setConstraints(constraints)
                .build()
            WorkManager.getInstance(context).enqueue(request)
            Timber.d("Scheduled auto-update");
        }

        fun updateNow(context: Context) {
            // TODO: Convert to coroutine scope
            // IBurnService(context).updateData() now returns suspend function
            Timber.d("updateNow called - needs coroutine implementation")
        }
    }

    override fun doWork(): Result {
        // A worker scheduled by another installed build can survive an app update.
        // Never let it replace a historical or mock build's bundled records with
        // whatever year the live endpoint currently serves.
        if (!BuildConfig.LIVE_DATA_UPDATES_ENABLED) {
            Timber.d("Skipping live data update for bundled annual data")
            return Result.success()
        }
        val dataProvider = DataProvider.getInstance(applicationContext)
        if (dataProvider.inUpgrade()) return Result.retry()

        val service = IBurnService(applicationContext)
        val success = try {
            runBlocking {
                service.updateData()
            }
        } catch (e: Exception) {
            Timber.e(e, "Update failed")
            false
        }
        Timber.d("Update task finished with success: $success")
        if (!success) {
            return Result.failure()
        }
        return Result.success()
    }
}
